arangod/RocksDBEngine/Listeners/RocksDBMetricsListener.cpp

```diff
@@ -0,0 +1,67 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2020 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Matthew Von-Maszewski
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBMetricsListener.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "RestServer/MetricsFeature.h"
+
+namespace arangodb {
+
+/// @brief Setup the object, clearing variables, but do no real work
+RocksDBMetricsListener::RocksDBMetricsListener(application_features::ApplicationServer& server)
+    : _writeStalls(server.getFeature<arangodb::MetricsFeature>().counter(
+          "rocksdb_write_stalls", 0,
+          "Number of times RocksDB has entered a stalled (slowed) write "
+          "state")),
+      _writeStops(server.getFeature<arangodb::MetricsFeature>().counter(
+          "rocksdb_write_stops", 0,
+          "Number of times RocksDB has entered a stopped write state")) {}
+
+void RocksDBMetricsListener::OnStallConditionsChanged(const rocksdb::WriteStallInfo& info) {
+  // we should only get here if there's an actual change
+  TRI_ASSERT(info.condition.cur != info.condition.prev);
+
+  // in the case that we go from normal to stalled or stopped, we count it;
+  // we also count a stall if we go from stopped to stall since it's a distinct
+  // state
+
+  if (info.condition.cur == rocksdb::WriteStallCondition::kDelayed) {
+    _writeStalls.count();
+    LOG_TOPIC("9123c", INFO, Logger::ENGINES)
+        << "rocksdb is slowing incoming writes to column family '"
+        << info.cf_name << "' to let background writes catch up";
```

Comment:
Would it make sense to have parameterized counters, so we have separate counters for each column family?

Reply:
That's a fair question. Personally I think seeing the counter increase and then checking the logs for more details is fine, but I'm open to making the change if we think it adds enough.

Reply:
Well, if we use parameterized counters here, then we should probably also use parameterized metrics for the time that spent in delayed mode.

Reply:
@jsteemann Care to weigh in?